File: nodejs/test/retry.test.ts
import { describe, expect, it } from 'vitest';
import { PactmanClient } from '../src/client.js';
import {
  PactmanAuthenticationError,
  PactmanRateLimitError,
  PactmanServerError,
  PactmanValidationError,
} from '../src/errors.js';
import { computeRetryDelay, readRetryAfter } from '../src/http.js';
import type { ResolvedRetryOptions } from '../src/config.js';
import {
  captureError,
  createClock,
  createFetchMock,
  envelope,
  nonprofitFixture,
  TEST_API_KEY,
  type ClockMock,
  type Stub,
} from './helpers.js';

const BASE_URL = 'http://mock.test';

function build(stubs: Stub[], overrides: Record<string, unknown> = {}, random = 1) {
  const mock = createFetchMock(stubs);
  const clock: ClockMock = createClock(random);
  const client = new PactmanClient(
    { apiKey: TEST_API_KEY, baseUrl: BASE_URL, fetch: mock.fetch, ...overrides },
    clock,
  );

  return { mock, clock, client };
}

const RETRY_DEFAULTS: ResolvedRetryOptions = {
  maxRetries: 2,
  initialDelayMs: 500,
  maxDelayMs: 8_000,
  backoffFactor: 2,
  jitter: false,
  retryableStatuses: [429, 500, 502, 503, 504],
  respectRetryAfter: true,
};

describe('automatic retries', () => {
  it('succeeds after a temporary failure and records the expected attempt count', async () => {
    const { mock, client } = build([
      { status: 503, body: { code: 503, message: 'Service Unavailable' } },
      { status: 200, body: envelope(nonprofitFixture()) },
    ]);

    const result = await client.nonprofits.check('411787097');

    expect(mock.requests).toHaveLength(2);
    expect(result.nonprofit?.ein).toBe('411787097');
  });

  it('never exceeds the configured maximum attempt count', async () => {
    const { mock, client } = build([{ status: 500, body: { code: 500 } }], {
      retry: { maxRetries: 3, jitter: false },
    });

    const error = await captureError<PactmanServerError>(client.nonprofits.check('411787097'));

    expect(mock.requests).toHaveLength(4);
    expect(error).toBeInstanceOf(PactmanServerError);
    expect(error.attempts).toBe(4);
  });

  it('makes a single attempt when retries are disabled', async () => {
    const { mock, client } = build([{ status: 500, body: { code: 500 } }], { retry: false });

    await client.nonprofits.check('411787097').catch(() => undefined);

    expect(mock.requests).toHaveLength(1);
  });

  it('retries temporary network failures', async () => {
    const { mock, client } = build([
      new TypeError('fetch failed'),
      { status: 200, body: envelope(nonprofitFixture()) },
    ]);

    await client.nonprofits.check('411787097');

    expect(mock.requests).toHaveLength(2);
  });

  it.each([
    ['401 authentication', 401],
    ['403 authorization', 403],
    ['400 validation', 400],
    ['404 not found', 404],
  ])('does not retry a %s response', async (_label, status) => {
    const { mock, client } = build([{ status, body: { code: status } }]);

    await client.nonprofits.check('411787097').catch(() => undefined);

    expect(mock.requests).toHaveLength(1);
  });

  it('surfaces a 401 as an authentication error on the first attempt', async () => {
    const { client } = build([{ status: 401, body: { code: 401 } }]);

    const error = await captureError<PactmanAuthenticationError>(
      client.nonprofits.check('411787097'),
    );

    expect(error).toBeInstanceOf(PactmanAuthenticationError);
    expect(error.attempts).toBe(1);
  });

  it('does not retry local validation errors', async () => {
    const { mock, client } = build([{ status: 200, body: envelope(nonprofitFixture()) }]);

    await expect(client.nonprofits.check('bad')).rejects.toBeInstanceOf(PactmanValidationError);
    expect(mock.requests).toHaveLength(0);
  });

  it('applies exponential backoff with a deterministic clock', async () => {
    const { clock, client } = build([{ status: 500, body: { code: 500 } }], {
      retry: { maxRetries: 3, jitter: false, initialDelayMs: 100, backoffFactor: 2 },
    });

    await client.nonprofits.check('411787097').catch(() => undefined);

    expect(clock.delays).toEqual([100, 200, 400]);
  });

  it('applies full jitter across the backoff window', async () => {
    const { clock, client } = build(
      [{ status: 500, body: { code: 500 } }],
      { retry: { maxRetries: 2, jitter: true, initialDelayMs: 100, backoffFactor: 2 } },
      0.5,
    );

    await client.nonprofits.check('411787097').catch(() => undefined);

    expect(clock.delays).toEqual([50, 100]);
  });

  it('caps a single backoff delay at maxDelayMs', async () => {
    const { clock, client } = build([{ status: 500, body: { code: 500 } }], {
      retry: { maxRetries: 3, jitter: false, initialDelayMs: 1_000, backoffFactor: 10, maxDelayMs: 2_000 },
    });

    await client.nonprofits.check('411787097').catch(() => undefined);

    expect(clock.delays).toEqual([1_000, 2_000, 2_000]);
  });

  it('accepts a per-request retry override', async () => {
    const { mock, client } = build([{ status: 500, body: { code: 500 } }], { retry: false });

    await client.nonprofits.check('411787097', { retry: { maxRetries: 1 } }).catch(() => undefined);

    expect(mock.requests).toHaveLength(2);
  });
});

describe('rate limiting', () => {
  it('maps 429 to the rate-limit error and exposes Retry-After', async () => {
    const { client } = build([{ status: 429, body: { code: 429 }, headers: { 'retry-after': '5' } }], {
      retry: false,
    });

    const error = await captureError<PactmanRateLimitError>(client.nonprofits.check('411787097'));

    expect(error).toBeInstanceOf(PactmanRateLimitError);
    expect(error.retryAfterSeconds).toBe(5);
  });

  it('waits for the server Retry-After before falling back to backoff', async () => {
    const { clock, client } = build(
      [
        { status: 429, body: { code: 429 }, headers: { 'retry-after': '7' } },
        { status: 200, body: envelope(nonprofitFixture()) },
      ],
      { retry: { maxRetries: 2, jitter: false, initialDelayMs: 100 } },
    );

    await client.nonprofits.check('411787097');

    expect(clock.delays).toEqual([7_000]);
  });

  it('ignores Retry-After when the caller opts out', async () => {
    const { clock, client } = build(
      [
        { status: 429, body: { code: 429 }, headers: { 'retry-after': '7' } },
        { status: 200, body: envelope(nonprofitFixture()) },
      ],
      { retry: { maxRetries: 2, jitter: false, initialDelayMs: 250, respectRetryAfter: false } },
    );

    await client.nonprofits.check('411787097');

    expect(clock.delays).toEqual([250]);
  });

  it('reads Retry-After given as an HTTP date', () => {
    const now = Date.UTC(2026, 0, 1, 12, 0, 0);
    const headers = new Headers({ 'retry-after': new Date(now + 4_000).toUTCString() });

    expect(readRetryAfter(headers, now)).toBe(4);
    expect(computeRetryDelay(1, RETRY_DEFAULTS, 4, () => 1)).toBe(4_000);
  });

  it('ignores an unparseable Retry-After', () => {
    expect(readRetryAfter(new Headers({ 'retry-after': 'soon' }))).toBeNull();
    expect(readRetryAfter(new Headers())).toBeNull();
  });

  it('spaces requests when a client-side limit is configured', async () => {
    const { clock, client } = build(
      [{ status: 200, body: envelope(nonprofitFixture()) }],
      { maxRequestsPerSecond: 2 },
    );

    await client.nonprofits.check('411787097');
    await client.nonprofits.check('411787097');
    await client.nonprofits.check('411787097');

    // Two requests per second means a 500ms slot each. The fake clock records
    // the wait without advancing wall time, so the schedule accumulates. Real
    // milliseconds still elapse between calls, so assert the window, not exact
    // values.
    expect(clock.delays).toHaveLength(2);
    expect(clock.delays[0]).toBeGreaterThan(400);
    expect(clock.delays[0]).toBeLessThanOrEqual(500);
    expect(clock.delays[1]).toBeGreaterThan(900);
    expect(clock.delays[1]).toBeLessThanOrEqual(1_000);
  });

  it('does not throttle when no client-side limit is set', async () => {
    const { clock, client } = build([{ status: 200, body: envelope(nonprofitFixture()) }]);

    await client.nonprofits.check('411787097');
    await client.nonprofits.check('411787097');

    expect(clock.delays).toEqual([]);
  });
});

describe('computeRetryDelay', () => {
  it('prefers a valid Retry-After over computed backoff', () => {
    expect(computeRetryDelay(1, RETRY_DEFAULTS, 3)).toBe(3_000);
  });

  it('falls back to backoff when Retry-After is absent', () => {
    expect(computeRetryDelay(1, RETRY_DEFAULTS, null)).toBe(500);
    expect(computeRetryDelay(3, RETRY_DEFAULTS, null)).toBe(2_000);
  });
});

File: ruby/test/retry_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class RetryTest < Minitest::Test
  include TestSupport

  def build(stubs, random: 1.0, advance: true, **)
    adapter = FakeAdapter.new(stubs)
    clock = FakeClock.new(random: random, advance: advance)

    [adapter, clock, client_with(adapter, clock: clock, **)]
  end

  def test_succeeds_after_a_temporary_failure
    adapter, _clock, client = build([{ status: 503, body: { "code" => 503 } }, { body: envelope(nonprofit_fixture) }])

    result = client.nonprofits.check("411787097")

    assert_equal 2, adapter.requests.size
    assert_equal "411787097", result.nonprofit.ein
  end

  def test_never_exceeds_the_configured_maximum_attempt_count
    adapter, _clock, client = build([{ status: 500, body: { "code" => 500 } }],
                                    retry: { max_retries: 3, jitter: false })

    error = assert_raises(NCP::ServerError) { client.nonprofits.check("411787097") }

    assert_equal 4, adapter.requests.size
    assert_equal 4, error.attempts
  end

  def test_makes_a_single_attempt_when_retries_are_disabled
    adapter, _clock, client = build([{ status: 500, body: { "code" => 500 } }], retry: false)

    assert_raises(NCP::ServerError) { client.nonprofits.check("411787097") }
    assert_equal 1, adapter.requests.size
  end

  def test_retries_temporary_network_failures
    adapter, _clock, client = build([Errno::ECONNRESET.new, { body: envelope(nonprofit_fixture) }])

    client.nonprofits.check("411787097")

    assert_equal 2, adapter.requests.size
  end

  def test_does_not_retry_authentication_authorization_validation_or_not_found
    [401, 403, 400, 404].each do |status|
      adapter, _clock, client = build([{ status: status, body: { "code" => status } }],
                                      retry: { retryable_statuses: [400, 401, 403, 404, 500] })

      assert_raises(NCP::ApiError) { client.nonprofits.check("411787097") }
      assert_equal 1, adapter.requests.size, "HTTP #{status} was retried"
    end
  end

  def test_surfaces_a_401_as_an_authentication_error_on_the_first_attempt
    _adapter, _clock, client = build([{ status: 401, body: { "code" => 401 } }])

    error = assert_raises(NCP::AuthenticationError) { client.nonprofits.check("411787097") }

    assert_equal 1, error.attempts
  end

  def test_does_not_retry_local_validation_errors
    adapter, _clock, client = build([{ body: envelope(nonprofit_fixture) }])

    assert_raises(NCP::ValidationError) { client.nonprofits.check("bad") }
    assert_empty adapter.requests
  end

  def test_applies_exponential_backoff_with_a_deterministic_clock
    _adapter, clock, client = build([{ status: 500, body: {} }],
                                    retry: { max_retries: 3, jitter: false, initial_delay: 0.1, backoff_factor: 2 })

    assert_raises(NCP::ServerError) { client.nonprofits.check("411787097") }
    assert_equal [0.1, 0.2, 0.4], clock.delays
  end

  def test_applies_full_jitter_across_the_backoff_window
    _adapter, clock, client = build([{ status: 500, body: {} }],
                                    random: 0.5, retry: { max_retries: 2, jitter: true, initial_delay: 0.1 })

    assert_raises(NCP::ServerError) { client.nonprofits.check("411787097") }
    assert_equal [0.05, 0.1], clock.delays
  end

  def test_caps_a_single_backoff_delay_at_max_delay
    _adapter, clock, client = build([{ status: 500, body: {} }],
                                    retry: { max_retries: 3, jitter: false, initial_delay: 1, backoff_factor: 10,
                                             max_delay: 2 })

    assert_raises(NCP::ServerError) { client.nonprofits.check("411787097") }
    assert_equal [1.0, 2.0, 2.0], clock.delays
  end

  def test_accepts_a_per_request_retry_override
    adapter, _clock, client = build([{ status: 500, body: {} }], retry: false)

    assert_raises(NCP::ServerError) { client.nonprofits.check("411787097", retry: { max_retries: 1 }) }
    assert_equal 2, adapter.requests.size
  end

  def test_accepts_a_per_request_opt_out
    adapter, _clock, client = build([{ status: 500, body: {} }])

    assert_raises(NCP::ServerError) { client.nonprofits.check_bulk(["411787097"], retry: false) }
    assert_equal 1, adapter.requests.size
  end

  def test_rejects_a_malformed_per_request_retry_before_sending
    adapter, _clock, client = build([{ body: envelope(nonprofit_fixture) }])

    assert_raises(NCP::ConfigurationError) { client.nonprofits.check("411787097", retry: { max_retries: -1 }) }
    assert_empty adapter.requests
  end
end

class RateLimitTest < Minitest::Test
  include TestSupport

  def test_maps_429_to_the_rate_limit_error_and_exposes_retry_after
    adapter = FakeAdapter.new([{ status: 429, body: { "code" => 429 }, headers: { "retry-after" => "5" } }])

    error = assert_raises(NCP::RateLimitError) { client_with(adapter, retry: false).nonprofits.check("411787097") }

    assert_equal 5, error.retry_after_seconds
  end

  def test_waits_for_the_server_retry_after_before_falling_back_to_backoff
    adapter = FakeAdapter.new([{ status: 429, body: {}, headers: { "retry-after" => "7" } },
                               { body: envelope(nonprofit_fixture) }])
    clock = FakeClock.new

    client_with(adapter, clock: clock, retry: { jitter: false, initial_delay: 0.1 }).nonprofits.check("411787097")

    assert_equal [7.0], clock.delays
  end

  def test_ignores_retry_after_when_the_caller_opts_out
    adapter = FakeAdapter.new([{ status: 429, body: {}, headers: { "retry-after" => "7" } },
                               { body: envelope(nonprofit_fixture) }])
    clock = FakeClock.new

    client_with(adapter, clock: clock, retry: { jitter: false, initial_delay: 0.25, respect_retry_after: false })
      .nonprofits.check("411787097")

    assert_equal [0.25], clock.delays
  end

  def test_reads_retry_after_given_as_an_http_date
    now = Time.utc(2026, 1, 1, 12, 0, 0)
    headers = { "retry-after" => (now + 4).httpdate }
    policy = NCP::RetryPolicy.new(jitter: false)

    assert_in_delta 4, NCP::Transport.read_retry_after(headers, now)
    assert_in_delta 4.0, NCP::Transport.compute_retry_delay(1, policy, 4, -> { 1 })
  end

  def test_treats_a_retry_after_date_in_the_past_as_zero
    now = Time.utc(2026, 1, 1, 12, 0, 0)

    assert_equal 0, NCP::Transport.read_retry_after({ "retry-after" => (now - 30).httpdate }, now)
  end

  def test_ignores_an_unparseable_or_negative_retry_after
    assert_nil NCP::Transport.read_retry_after({ "retry-after" => "soon" })
    assert_nil NCP::Transport.read_retry_after({ "retry-after" => "-5" })
    assert_nil NCP::Transport.read_retry_after({ "retry-after" => "  " })
    assert_nil NCP::Transport.read_retry_after({})
  end

  def test_spaces_requests_when_a_client_side_limit_is_configured
    adapter = FakeAdapter.new([{ body: envelope(nonprofit_fixture) }])
    clock = FakeClock.new
    client = client_with(adapter, clock: clock, max_requests_per_second: 2)

    3.times { client.nonprofits.check("411787097") }

    # Two requests per second is a half-second slot each. The first goes out at
    # once; each after it waits for the slot the previous one reserved.
    assert_equal [0.5, 0.5], clock.delays
  end

  def test_accumulates_the_schedule_for_a_burst
    adapter = FakeAdapter.new([{ body: envelope(nonprofit_fixture) }])
    clock = FakeClock.new(advance: false)
    client = client_with(adapter, clock: clock, max_requests_per_second: 2)

    3.times { client.nonprofits.check("411787097") }

    # Time stands still, so the third request queues behind the second.
    assert_equal [0.5, 1.0], clock.delays
  end

  def test_shares_one_schedule_across_threads
    adapter = FakeAdapter.new([{ body: envelope(nonprofit_fixture) }])
    clock = FakeClock.new(advance: false)
    client = client_with(adapter, clock: clock, max_requests_per_second: 4)

    Array.new(4) { Thread.new { client.nonprofits.check("411787097") } }.each(&:join)

    assert_equal [0.25, 0.5, 0.75], clock.delays.sort
  end

  def test_does_not_throttle_when_no_client_side_limit_is_set
    adapter = FakeAdapter.new([{ body: envelope(nonprofit_fixture) }])
    clock = FakeClock.new
    client = client_with(adapter, clock: clock)

    2.times { client.nonprofits.check("411787097") }

    assert_empty clock.delays
  end
end

class ComputeRetryDelayTest < Minitest::Test
  include TestSupport

  def test_prefers_a_valid_retry_after_over_computed_backoff
    assert_in_delta 3.0, NCP::Transport.compute_retry_delay(1, NCP::RetryPolicy.new(jitter: false), 3)
  end

  def test_falls_back_to_backoff_when_retry_after_is_absent
    policy = NCP::RetryPolicy.new(jitter: false)

    assert_in_delta 0.5, NCP::Transport.compute_retry_delay(1, policy, nil)
    assert_in_delta 2.0, NCP::Transport.compute_retry_delay(3, policy, nil)
  end
end
